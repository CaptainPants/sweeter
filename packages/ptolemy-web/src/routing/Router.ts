import { $derived, Component } from '@serpentis/ptolemy-core';

import { getPath } from './implementation/getPath.js';
import { pathDoesNotMatch } from './pathDoesNotMatch.js';
import { type Route } from './types.js';

export interface RouterProps {
    basePath?: string;
    url: string;

    routes: readonly Route<readonly string[]>[];

    fallback?: (() => JSX.Element) | undefined;
}

export const Router: Component<RouterProps> = (
    { basePath, url, routes, fallback },
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    init,
) => {
    const asUrl = $derived(() => new URL(url.value));

    return $derived(() => {
        // Allow for '/' as a base path
        let basePathResolved = basePath?.value ?? '';
        if (basePathResolved.startsWith('/')) {
            basePathResolved = basePathResolved.substring(1);
        }

        const path = getPath(asUrl.value, basePathResolved);

        if (path !== null) {
            for (const route of routes.value) {
                // basePath should be removed from path
                const match = route.match(path, asUrl.value);

                if (match !== pathDoesNotMatch) {
                    return match;
                }
            }
        }

        return fallback?.value?.();
    });
};
