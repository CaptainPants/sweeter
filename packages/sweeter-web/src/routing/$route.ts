import { pathDoesNotMatch } from './pathDoesNotMatch.js';
import { type PathTemplate, type Route } from './types.js';

export interface RouteRenderFunction<TArguments extends readonly string[]> {
    (args: TArguments, path: string, url: URL): JSX.Element;
}

export interface RouteHandler<TArguments extends readonly string[]> {
    render: RouteRenderFunction<TArguments>;
}

export function $route<const TArguments extends readonly string[]>(
    pathTemplate: PathTemplate<TArguments>,
    setup: () => RouteHandler<TArguments> | RouteRenderFunction<TArguments>,
): Route<readonly string[]> {
    const setupResult = setup();
    const render =
        typeof setupResult === 'function' ? setupResult : setupResult.render;

    return {
        pathTemplate: pathTemplate,
        match: (path, url) => {
            const match = pathTemplate.match(path);
            if (match) {
                return render(match, path, url);
            }

            return pathDoesNotMatch;
        },
    };
}
