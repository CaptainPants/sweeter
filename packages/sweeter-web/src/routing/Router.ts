import {
    type ComponentInit,
    type PropertiesMightBeSignals,
} from '@captainpants/sweeter-core';
import { $derived, $val } from '@captainpants/sweeter-core';

import { getPath } from './implementation/getPath.js';
import { pathDoesNotMatch } from './pathDoesNotMatch.js';
import { type Route } from './types.js';

export interface RouterProps {
    basePath?: string;
    url: string;

    routes: readonly Route<readonly string[]>[];

    fallback?: (() => JSX.Element) | undefined;
}

export function Router(
    { basePath, url, routes, fallback }: PropertiesMightBeSignals<RouterProps>,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    init: ComponentInit,
): JSX.Element {
    const asUrl = $derived(() => new URL($val(url)));

    return $derived(() => {
        // Allow for '/' as a base path
        let basePathResolved = $val(basePath) ?? "";
        if (basePathResolved.startsWith('/')) {
            basePathResolved = basePathResolved.substring(1);
        }

        const path = getPath(asUrl.value, basePathResolved);

        if (path !== null) {
            for (const route of $val(routes)) {
                // basePath should be removed from path
                const match = route.match(path, asUrl.value);

                if (match !== pathDoesNotMatch) {
                    return match;
                }
            }
        }

        return $val(fallback)?.();
    });
}
