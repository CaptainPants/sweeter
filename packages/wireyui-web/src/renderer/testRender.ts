import { WebRenderer } from './WebRenderer.js';

export interface TestRenderResult {
    readonly nodes: NodeListOf<ChildNode>;
    readonly getHTML: () => string;
    readonly dispose: () => void;
}

/**
 * Convenience function for rendering isolated JSX chunks for unit tests
 * @param callback
 * @returns
 */
export function testRender(callback: () => JSX.Element): TestRenderResult {
    const renderer = new WebRenderer();

    const root = document.createElement('div');

    const dispose = renderer.start(root, callback);

    return {
        nodes: root.childNodes,
        getHTML: () => root.innerHTML,
        dispose,
    };
}
