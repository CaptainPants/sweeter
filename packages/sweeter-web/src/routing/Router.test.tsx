/* @jsxImportSource .. */

import {
    $lazyComponent,
    $mutable,
    Suspense,
    type Component,
} from '@captainpants/sweeter-core';
import { testRender } from '../test/testRender.js';
import { Router } from './Router.js';
import { match, pathTemplate } from './index.js';
import { $route } from './$route.js';

it('General', async () => {
    const routes = [
        $route(pathTemplate`this/is/a/${match.segment}`, () => {
            const Component = $lazyComponent(() =>
                Promise.resolve<Component<{ text: string }>>((props) => {
                    return `Text: ${props.text}`;
                }),
            );

            return ([arg], path, url) => {
                return <Component text="Test" />;
            };
        }),
    ];

    const path = $mutable('this/is/a/banana');

    const res = testRender(() => (
        <Suspense fallback={() => 'Loading'}>
            {() => (
                <Router
                    routes={routes}
                    rootRelativePath={path}
                    url={new URL('https://google.com/this/is/a/banana')}
                />
            )}
        </Suspense>
    ));

    // LOADING
    expect(res.getHTML()).toMatchSnapshot();

    await new Promise((resolve) => queueMicrotask(() => resolve(void 0)));

    // LOADED
    expect(res.getHTML()).toMatchSnapshot();

    res.dispose();
});
