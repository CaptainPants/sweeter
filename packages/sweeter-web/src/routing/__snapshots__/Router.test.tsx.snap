// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`General 1`] = `"Loading<div style=\\"display: none;\\"><!--Hooks(Async): subscribeToChanges--></div>"`;

exports[`General 2`] = `"Text: Test<!--Hooks(Async): subscribeToChanges-->"`;
